pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,399 @@
+		return
+	}
+	if !exists {
+		http.Error(w, fmt.Sprintf("upload of block %s not started yet", blockID), http.StatusNotFound)
+		return
+	}
+
+	// TODO: Verify that upload path and length correspond to file index
+
+	dst := path.Join(blockID, pth)
+
+	level.Debug(logger).Log("msg", "uploading block file to bucket", "destination", dst,
+		"size", r.ContentLength)
+	reader := bodyReader{
+		r: r,
+	}
+	if err := userBkt.Upload(ctx, dst, reader); err != nil {
+		level.Error(logger).Log("msg", "failed uploading block file to bucket",
+			"destination", dst, "err", err)
+		// We don't know what caused the error; it could be the client's fault (e.g. killed
+		// connection), but internal server error is the safe choice here.
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+		return
+	}
+
+	level.Debug(logger).Log("msg", "finished uploading block file to bucket",
+		"path", pth)
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func decodeMeta(r io.Reader, name string) (metadata.Meta, error) {
+	dec := json.NewDecoder(r)
+	var meta metadata.Meta
+	if err := dec.Decode(&meta); err != nil {
+		return meta, errors.Wrap(err, fmt.Sprintf("failed decoding %s", name))
+	}
+
+	return meta, nil
+}
+
+func (c *MultitenantCompactor) completeBlockUpload(ctx context.Context, r *http.Request,
+	logger log.Logger, userBkt objstore.Bucket, tenantID string, blockID ulid.ULID) error {
+	level.Debug(logger).Log("msg", "received request to complete block upload", "content_length", r.ContentLength)
+
+	tmpMetaPath := path.Join(blockID.String(), tmpMetaFilename)
+	rdr, err := userBkt.Get(ctx, tmpMetaPath)
+	if err != nil {
+		if userBkt.IsObjNotFoundErr(err) {
+			return httpError{
+				message:    fmt.Sprintf("upload of block %s not started yet", blockID),
+				statusCode: http.StatusNotFound,
+			}
+		}
+		return errors.Wrap(err, fmt.Sprintf("failed to download %s from object storage", tmpMetaFilename))
+	}
+	defer func() {
+		_ = rdr.Close()
+	}()
+
+	meta, err := decodeMeta(rdr, tmpMetaFilename)
+	if err != nil {
+		return err
+	}
+
+	level.Debug(logger).Log("msg", "completing block upload", "files", len(meta.Thanos.Files))
+
+	// Upload meta file so block is considered complete
+	if err := c.uploadMeta(ctx, logger, meta, blockID, block.MetaFilename, userBkt); err != nil {
+		return err
+	}
+
+	if err := userBkt.Delete(ctx, path.Join(blockID.String(), tmpMetaFilename)); err != nil {
+		level.Warn(logger).Log("msg", fmt.Sprintf(
+			"failed to delete %s from block in object storage", tmpMetaFilename), "err", err)
+		return nil
+	}
+
+	level.Debug(logger).Log("msg", "successfully completed block upload")
+	return nil
+}
+
+// sanitizeMeta sanitizes and validates a metadata.Meta object. If a validation error occurs, an error
+// message gets returned, otherwise an empty string.
+func (c *MultitenantCompactor) sanitizeMeta(logger log.Logger, tenantID string, blockID ulid.ULID,
+	meta *metadata.Meta) string {
```

Comment:
Since we don't do anything with meta.json later, we should do some more sanitization:
- check if Version == 1
- remove Compaction.Parents
- set Compaction.Sources to the block itself

Reply:
Done, thanks!